src/hotspot/share/opto/parse2.cpp

```diff
@@ -373,178 +373,215 @@ Node* Parse::array_addressing(BasicType type, int vals, const Type*& elemtype) {
-        uncommon_trap(Deoptimization::Reason_range_check,
-                      Deoptimization::Action_make_not_entrant,
-                      nullptr, "range_check");
-      } else {
-        // If we have already recompiled with the range-check-widening
-        // heroic optimization turned off, then we must really be throwing
-        // range check exceptions.
-        builtin_throw(Deoptimization::Reason_range_check);
-      }
-    }
+  if (need_range_check) {
+    create_range_check(idx, ary, sizetype);
   }
+
   // Check for always knowing you are throwing a range-check exception
   if (stopped())  return top();
 
-  // This could be an access to an inline type array. We can't tell if it's
-  // flat or not. Knowing the exact type avoids runtime checks and leads to
-  // a much simpler graph shape. Check profile information.
-  if (!arytype->is_flat() && !arytype->is_not_flat()) {
-    // First check the speculative type
-    Deoptimization::DeoptReason reason = Deoptimization::Reason_speculate_class_check;
-    ciKlass* array_type = arytype->speculative_type();
-    if (too_many_traps_or_recompiles(reason) || array_type == nullptr) {
-      // No speculative type, check profile data at this bci
-      array_type = nullptr;
-      reason = Deoptimization::Reason_class_check;
-      if (UseArrayLoadStoreProfile && !too_many_traps_or_recompiles(reason)) {
-        ciKlass* element_type = nullptr;
-        ProfilePtrKind element_ptr = ProfileMaybeNull;
-        bool flat_array = true;
-        bool null_free_array = true;
-        method()->array_access_profiled_type(bci(), array_type, element_type, element_ptr, flat_array, null_free_array);
-      }
-    }
-    if (array_type != nullptr) {
-      // Speculate that this array has the exact type reported by profile data
-      Node* better_ary = nullptr;
-      DEBUG_ONLY(Node* old_control = control();)
-      Node* slow_ctl = type_check_receiver(ary, array_type, 1.0, &better_ary);
-      if (stopped()) {
-        // The check always fails and therefore profile information is incorrect. Don't use it.
-        assert(old_control == slow_ctl, "type check should have been removed");
-        set_control(slow_ctl);
-      } else if (!slow_ctl->is_top()) {
-        { PreserveJVMState pjvms(this);
-          set_control(slow_ctl);
-          uncommon_trap_exact(reason, Deoptimization::Action_maybe_recompile);
-        }
-        replace_in_map(ary, better_ary);
-        ary = better_ary;
-        arytype  = _gvn.type(ary)->is_aryptr();
-        elemtype = arytype->elem();
-      }
+  // Make array address computation control dependent to prevent it
+  // from floating above the range check during loop optimizations.
+  Node* ptr = array_element_address(ary, idx, type, sizetype, control());
+  assert(ptr != top(), "top should go hand-in-hand with stopped");
+
+  return ptr;
+}
+
+void Parse::create_range_check(Node* idx, Node* ary, const TypeInt* sizetype) {
```

Comment:
Only extracted the code to a separate method without applying any refactoring to it.